File: backend/models/ScanResult.js
const mongoose = require('mongoose');

const scanResultSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  imagePath: {
    type: String,
    required: true
  },
  imageUrl: {
    type: String,
    required: true
  },
  prediction: {
    type: String,
    required: true
  },
  confidence: {
    type: Number,
    required: true,
    min: 0,
    max: 1
  },
  riskLevel: {
    type: String,
    required: true,
    enum: ['Low', 'Medium', 'High']
  },
  recommendations: [{
    type: String
  }],
  additionalData: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  scanDate: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Index for efficient queries
scanResultSchema.index({ userId: 1, scanDate: -1 });
scanResultSchema.index({ prediction: 1 });
scanResultSchema.index({ riskLevel: 1 });

// Virtual for confidence percentage
scanResultSchema.virtual('confidencePercentage').get(function() {
  return Math.round(this.confidence * 100);
});

// Method to get risk level based on confidence and prediction
scanResultSchema.methods.calculateRiskLevel = function() {
  if (this.prediction === 'Malignant') {
    return this.confidence > 0.7 ? 'High' : 'Medium';
  } else if (this.prediction === 'Suspicious') {
    return this.confidence > 0.6 ? 'Medium' : 'Low';
  } else {
    return 'Low';
  }
};

// Static method to get user statistics
scanResultSchema.statics.getUserStats = async function(userId) {
  const stats = await this.aggregate([
    { $match: { userId: new mongoose.Types.ObjectId(userId) } },
    {
      $group: {
        _id: null,
        totalScans: { $sum: 1 },
        highRisk: {
          $sum: { $cond: [{ $eq: ['$riskLevel', 'High'] }, 1, 0] }
        },
        mediumRisk: {
          $sum: { $cond: [{ $eq: ['$riskLevel', 'Medium'] }, 1, 0] }
        },
        lowRisk: {
          $sum: { $cond: [{ $eq: ['$riskLevel', 'Low'] }, 1, 0] }
        },
        avgConfidence: { $avg: '$confidence' },
        lastScan: { $max: '$scanDate' }
      }
    }
  ]);
  
  return stats.length > 0 ? stats[0] : {
    totalScans: 0,
    highRisk: 0,
    mediumRisk: 0,
    lowRisk: 0,
    avgConfidence: 0,
    lastScan: null
  };
};

module.exports = mongoose.model('ScanResult', scanResultSchema);

File: lib/models/scan_result.dart
import '../constants/app_constants.dart';

class ScanResult {
  final String id;
  final String userId;
  final String imagePath;
  final String imageUrl;
  final double confidence;
  final String prediction;
  final String riskLevel;
  final List<String> recommendations;
  final DateTime scanDate;
  final Map<String, dynamic>? additionalData;

  ScanResult({
    required this.id,
    required this.userId,
    required this.imagePath,
    required this.imageUrl,
    required this.confidence,
    required this.prediction,
    required this.riskLevel,
    required this.recommendations,
    required this.scanDate,
    this.additionalData,
  });

  factory ScanResult.fromJson(Map<String, dynamic> json) {
    return ScanResult(
      id: json['id'] ?? json['_id'],
      userId: json['userId'],
      imagePath: json['imagePath'],
      imageUrl: json['imageUrl'],
      confidence: json['confidence'].toDouble(),
      prediction: json['prediction'],
      riskLevel: json['riskLevel'],
      recommendations: List<String>.from(json['recommendations']),
      scanDate: DateTime.parse(json['scanDate']),
      additionalData: json['additionalData'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'imagePath': imagePath,
      'imageUrl': imageUrl,
      'confidence': confidence,
      'prediction': prediction,
      'riskLevel': riskLevel,
      'recommendations': recommendations,
      'scanDate': scanDate.toIso8601String(),
      'additionalData': additionalData,
    };
  }

  String get confidencePercentage => '${(confidence * 100).toStringAsFixed(1)}%';
  
  String get riskLevelString {
    switch (riskLevel.toLowerCase()) {
      case 'low':
        return 'Low Risk';
      case 'medium':
        return 'Medium Risk';
      case 'high':
        return 'High Risk';
      default:
        return 'Unknown Risk';
    }
  }
  
  RiskLevel get riskLevelEnum => RiskLevelExtension.fromString(riskLevel);
  
  bool get isHighRisk => riskLevel.toLowerCase() == 'high';
  bool get isMediumRisk => riskLevel.toLowerCase() == 'medium';
  bool get isLowRisk => riskLevel.toLowerCase() == 'low';
}